pkg/pillar/hypervisor/xen.go

```diff
@@ -829,3 +833,50 @@ func execWithTimeout(command string, args ...string) ([]byte, bool, error) {
 	}
 	return out, true, err
 }
+
+//verifyDomainState: Verifies if a domain is up and running by its state. Domain can have the following states:
+//r - currently running
+//b - blocked, and not running or runnable
+//p - paused
+//s - a shutdown command has been sent, but the domain isn't dying yet
+//c - the domain has crashed
+//d - the domain is dying, but hasn't properly shut down or crashed
+func verifyDomainState(domainName string) error {
+	for i := 2; i > 0; i-- {
+		//getStateCmd: command to filter domain state from 'xl list <domain>' output
+		getStateCmd := fmt.Sprintf("bash -c 'xl list %s' | awk '{print $5}' | awk 'NR==2'", domainName)
+		cmd := "sh"
+		args := []string{
+			"-c",
+			getStateCmd,
+		}
+		stdoutStderr, err := wrap.Command(cmd, args...).CombinedOutput()
+		if err != nil {
+			//domain is not present
+			log.Errorln("verifyDomainState: xl list failed ", err)
+			log.Errorln("verifyDomainState: xl list output ", string(stdoutStderr))
+			return fmt.Errorf("verifyDomainState: xl list failed: %s\n",
+				string(stdoutStderr))
+		}
+		log.Infof("verifyDomainState: domain: %s state: %s", domainName, stdoutStderr)
+		//Removing non set bits from domain state which are represented by "-"
+		domainState := strings.TrimSpace(strings.Replace(string(stdoutStderr), "-", "", -1))
+		if len(domainState) > 1 {
+			//Sometimes the domain can be in multiple states (logically possible) at the same time.
+			// In such cases we can consider the last state. For example when a domain is shutting down its state will be
+			//"--psc-", in this case we can consider "c" to represent that domain is in middle of shutting down.
+			domainState = domainState[len(domainState)-1:]
+		}
+		log.Infof("verifyDomainState: domain: %s domainState: %s.", domainName, domainState)
+		if domainState == "r" || domainState == "b" {
+			//domain is up and running.
+			return nil
+		} else if i > 0 { //No need to wait on last iteration.
+			//If domain is in the middle of shutting down, wait for 3 sec and check again.
+			time.Sleep(3 * time.Second)
```

Comment:
Can't you just assume that is it shutting down here and not check for the qemu process?

Basically make this function into isDomainShuttingDown() bool, and if that is the case don't check for qemu.
In that case you shouldn't have to depend on sleep.

Reply:
Changed.